### ISSUE TITLE
Incomplete Needed is incomplete when used in chain!

### ISSUE THREAD

**Issue by johansigfrids** [Geal/nom#94]
Chaining parser seems to cause a issue with `IResult::Incomplete(Needed::Size(n))`. If you have a `chain!` parser like:
```rust
chain!(input,
  a: take!(4) ~
  b: take!(8),
  ||{(a,b )}
)
```
and you start out with an input of length 8, then the first `take!` will successfully take 4 bytes, but the second `take!` will fail with a result `IResult::Incomplete(Needed::Size(8))` because there are only four bytes left and it needs eight. If this is then navily turned into a `ConsumerState::Await(0, 8)` the consumer will think everything is fine since there are 8 bytes available and so will attempt to run the parser again. This will lead to an endless loop. 

The problem is that when reporting on the needed bytes the previous bytes taken by `take!(4)` aren't taken into account. What is really needed is 12 bytes, but this is not reported in the `IResult::Incomplete(Needed::Size(n))` result.

**Comment by Geal** [comment]
I see. That bug is annoying, because `chain!` has to be completely generic with its input, so it cannot mess with slice size.

Maybe the solution would be to add a `More(usize)` variant to `Needed`. Or I can have some trait to compare input types.

Let me think about it for a bit, then I'll get back to you.

Thank you for catching that bug!

**Comment by Geal** [comment]
This is fixed by b1e1560 and c1ea46a. In the end, I did not need to change the definition of `Needed`. Now, `chain!` accumulates how much data was consumed. The issue is that it requires something that implements `len()` as input, but this could be abstracted away.

**Comment by Geal** [comment]
Fixed in 1.0
